File: examples/python/mock_device.py
#!/usr/bin/env python3
"""Example of SHDLC communication using virtual in-memory mock port."""

from rust_shdlc_driver import ShdlcConnection, ShdlcDevice, ShdlcMockPort


def make_resp(addr: int, cmd: int, state: int, data: bytes) -> bytes:
    content = bytearray([addr, cmd, state, len(data)]) + bytearray(data)
    checksum = (~sum(content)) & 0xFF
    content.append(checksum)
    stuffed = bytearray()
    for b in content:
        if b in [0x7E, 0x7D, 0x11, 0x13]:
            stuffed.append(0x7D)
            stuffed.append(b ^ 0x20)
        else:
            stuffed.append(b)
    return bytes(bytearray([0x7E]) + stuffed + bytearray([0x7E]))


def main():
    # Create mock port and push a simulated SHDLC response for GetProductName
    port = ShdlcMockPort(bitrate=115200)
    port.push_rx_data(make_resp(1, 0xD0, 0x00, b"SHT31\0"))

    conn = ShdlcConnection(port)
    device = ShdlcDevice(conn, slave_address=1)

    name = device.get_product_name()
    print(f"Mock device product name: {name}")
    assert name == "SHT31"


if __name__ == "__main__":
    main()

File: examples/mock_device.rs
use rust_shdlc_driver::connection::ShdlcConnection;
use rust_shdlc_driver::device::ShdlcDevice;
use rust_shdlc_driver::protocol::{calculate_checksum, stuff_data_bytes, START_STOP_BYTE};
use rust_shdlc_driver::transport::MockTransport;

fn make_response_frame(slave_addr: u8, cmd_id: u8, state: u8, data: &[u8]) -> Vec<u8> {
    let mut content = vec![slave_addr, cmd_id, state, data.len() as u8];
    content.extend_from_slice(data);
    let checksum = calculate_checksum(&content);
    content.push(checksum);
    let stuffed = stuff_data_bytes(&content);
    let mut raw = vec![START_STOP_BYTE];
    raw.extend_from_slice(&stuffed);
    raw.push(START_STOP_BYTE);
    raw
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let mock = MockTransport::new();
    // Simulate a device at address 0x00 responding to GetProductName (0xD0) with "SHT31\0"
    mock.push_rx_data(&make_response_frame(0x00, 0xD0, 0x00, b"SHT31\0"));

    let conn = ShdlcConnection::new(Box::new(mock))?;
    let mut device = ShdlcDevice::new(conn, 0);

    let product_name = device.get_product_name()?;
    println!("Device responded with product name: {}", product_name);
    assert_eq!(product_name, "SHT31");

    Ok(())
}
